proposals/text/0000-shaper-in-prompt-template.md

```diff
@@ -0,0 +1,153 @@
+ 'root_node': 'Query',
+ 'params': {},
+ 'query': 'What is the most popular drink?',
+ 'documents': [<Document: {'content': 'Beer is the oldest[1][2][3] and most widely consumed[4] type of alcoholic drink in the world, and the third most popular drink overall after potable water and tea.[5] It is produced by the brewing and fermentation of starches, mainly derived from cereal grains—most commonly from malted barley, though wheat, maize (corn), rice, and oats are also used. During the brewing process, fermentation of the starch sugars in the wort produces ethanol and carbonation in the resulting beer.[6] Most modern beer is brewed with hops, which add bitterness and other flavours and act as a natural preservative and stabilizing agent. Other flavouring agents such as gruit, herbs, or fruits may be included or used instead of hops. In commercial brewing, the natural carbonation effect is often removed during processing and replaced with forced carbonation.[7]', 'content_type': 'text', 'score': None, 'meta': {}, 'id_hash_keys': ['content'], 'embedding': None, 'id': 'fcd62336fb380a69c2d655f8cd072995'}>],
+ 'node_id': 'PromptNode'}
+    ```
+
+The corresponding prompt template would look like this (provided `value_to_list` and `join_documents` Shaper functions are extended a bit):
+
+    ```python
+    PromptTemplate(
+            name="question-answering-with-references",
+            prompt_text="Create a concise and informative answer (no more than 50 words) for a given question "
+            "based solely on the given documents. You must only use information from the given documents. "
+            "Use an unbiased and journalistic tone. Do not repeat text. Cite the documents using Document[number] notation. "
+            "If multiple documents contain the answer, cite those documents like ‘as stated in Document[number,number,etc]’. "
+            "If the documents do not contain the answer to the question, say that ‘answering is not possible given the available information.’\n"
+            "$documents \n Question: $query; Answer: ",
+            input_shapers=[
+                Shaper(
+                    func="join_documents",
+                    inputs={
+                        "documents": "documents",
+                        "delimiter": "delimiter",
+                        "pattern": "pattern",
+                        "str_replace": "str_replace",
+                    },
+                    outputs=["documents"],
+                    params={
+                        "delimiter": "\n",
+                        "pattern": "\nDocument[$idx]: $content",
```

Comment:
For me, the valid `$` variables in the `pattern` variable are hard to determine without looking at the source code of the `join_documents` function. I'd recommend adding docstrings for what the valid `$` variables are and maybe provide checks in the function if an invalid `pattern` are passed.